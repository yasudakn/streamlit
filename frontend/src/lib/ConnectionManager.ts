/**
 * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { BackMsg, ForwardMsg } from "src/autogen/proto"
import { BaseUriParts, getPossibleBaseUris } from "src/lib/UriUtil"
import { ReactNode } from "react"

import { ConnectionState } from "./ConnectionState"
import { logError } from "./log"
import { WebsocketConnection } from "./WebsocketConnection"
import { ensureError } from "./ErrorHandling"

/**
 * When the websocket connection retries this many times, we show a dialog
 * letting the user know we're having problems connecting.
 */
const RETRY_COUNT_FOR_WARNING = 30 // around 15s

interface Props {
  /**
   * Function to be called when we receive a message from the server.
   */
  onMessage: (message: ForwardMsg) => void

  /**
   * Function to be called when the connection errors out.
   */
  onConnectionError: (errNode: ReactNode) => void

  /**
   * Called when our ConnectionState is changed.
   */
  connectionStateChanged: (connectionState: ConnectionState) => void
}

/**
 * Manages our connection to the Server.
 */
export class ConnectionManager {
  private readonly props: Props

  private connection?: WebsocketConnection

  private connectionState: ConnectionState = ConnectionState.INITIAL

  constructor(props: Props) {
    this.props = props

    // This method returns a promise, but we don't care about its result.
    this.connect()
  }

  /**
   * Indicates whether we're connected to the server.
   */
  public isConnected(): boolean {
    return this.connectionState === ConnectionState.CONNECTED
  }

  /**
   * Return the BaseUriParts for the server we're connected to,
   * if we are connected to a server.
   */
  public getBaseUriParts(): BaseUriParts | undefined {
    if (this.connection instanceof WebsocketConnection) {
      return this.connection.getBaseUriParts()
    }
    return undefined
  }

  public sendMessage(obj: BackMsg): void {
    if (this.connection instanceof WebsocketConnection && this.isConnected()) {
      this.connection.sendMessage(obj)
    } else {
      // Don't need to make a big deal out of this. Just print to console.
      logError(`Cannot send message when server is disconnected: ${obj}`)
    }
  }

  /**
   * Increment the runCount on our message cache, and clear entries
   * whose age is greater than the max.
   */
  public incrementMessageCacheRunCount(maxMessageAge: number): void {
    // StaticConnection does not use a MessageCache.
    if (this.connection instanceof WebsocketConnection) {
      this.connection.incrementMessageCacheRunCount(maxMessageAge)
    }
  }

  private async connect(): Promise<void> {
    try {
      this.connection = await this.connectToRunningServer()
    } catch (e) {
      const err = ensureError(e)
      logError(err.message)
      this.setConnectionState(
        ConnectionState.DISCONNECTED_FOREVER,
        err.message
      )
    }
  }

  private setConnectionState = (
    connectionState: ConnectionState,
    errMsg?: string
  ): void => {
    if (this.connectionState !== connectionState) {
      this.connectionState = connectionState
      this.props.connectionStateChanged(connectionState)
    }

    if (errMsg || connectionState === ConnectionState.DISCONNECTED_FOREVER) {
      this.props.onConnectionError(errMsg || "unknown")
    }
  }

  private showRetryError = (
    totalRetries: number,
    latestError: ReactNode
  ): void => {
    if (totalRetries === RETRY_COUNT_FOR_WARNING) {
      this.props.onConnectionError(latestError)
    }
  }

  private connectToRunningServer(): WebsocketConnection {
    const baseUriPartsList = getPossibleBaseUris()

    return new WebsocketConnection({
      baseUriPartsList,
      onMessage: this.props.onMessage,
      onConnectionStateChange: this.setConnectionState,
      onRetry: this.showRetryError,
    })
  }
}
