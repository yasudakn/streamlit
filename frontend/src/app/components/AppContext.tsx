/**
 * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from "react"

import { PageConfig } from "src/lib/proto"

export interface Props {
  /**
   * If true, render the app with a wider column size.
   * Set from the UserSettings object.
   * @see UserSettings
   */
  wideMode: boolean

  /**
   * The sidebar's default display state.
   * Set from the PageConfig protobuf.
   */
  initialSidebarState: PageConfig.SidebarState

  /**
   * True if the app is embedded.
   * @see isEmbed
   */
  embedded: boolean

  /**
   * True if padding is enabled.
   * @see isPaddingDisplayed
   */
  showPadding: boolean

  /**
   * True if scrolling is disabled.
   * @see isScrollingHidden
   */
  disableScrolling: boolean

  /**
   * True if the footer should be displayed.
   * @see isFooterDisplayed
   */
  showFooter: boolean

  /**
   * True if the toolbar should be displayed.
   * @see isToolbarDisplayed
   */
  showToolbar: boolean

  /**
   * True if the thin colored line at the top of the app should be displayed.
   * @see isColoredLineDisplayed
   */
  showColoredLine: boolean

  /**
   * Part of URL construction for an app page in a multi-page app;
   * this is set from the host communication manager via host message.
   * @see SidebarNav
   */
  pageLinkBaseUrl: string

  /**
   * If non-zero, this is the number of pixels that the sidebar's
   * "chevron" icon is shifted. (If sidebarChevronDownshift is 0, then
   * the current theme's spacing is used.);
   * this is set from the host communication manager via host message.
   * @see StyledSidebarCollapsedControl
   */
  sidebarChevronDownshift: number
}

export const AppContext = React.createContext<Props>({
  wideMode: false,
  initialSidebarState: PageConfig.SidebarState.AUTO,
  embedded: false,
  showPadding: false,
  disableScrolling: false,
  showFooter: false,
  showToolbar: false,
  showColoredLine: false,
  pageLinkBaseUrl: "",
  sidebarChevronDownshift: 0,
})
