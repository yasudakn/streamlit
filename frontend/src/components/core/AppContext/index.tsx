/**
 * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from "react"

import { PageConfig } from "src/autogen/proto"
import { baseTheme, ThemeConfig } from "src/theme"
import { BaseUriParts, getWindowBaseUriParts } from "src/lib/UriUtil"

export interface Props {
  wideMode: boolean
  layout: PageConfig.Layout
  initialSidebarState: PageConfig.SidebarState
  embedded: boolean
  isFullScreen: boolean
  setFullScreen: (value: boolean) => void
  addScriptFinishedHandler: (func: () => void) => void
  removeScriptFinishedHandler: (func: () => void) => void
  activeTheme: ThemeConfig
  setTheme: (theme: ThemeConfig) => void
  availableThemes: ThemeConfig[]
  addThemes: (themes: ThemeConfig[]) => void
  sidebarChevronDownshift: number
  getBaseUriParts: () => BaseUriParts | undefined
}

export default React.createContext<Props>({
  wideMode: false,
  layout: PageConfig.Layout.CENTERED,
  initialSidebarState: PageConfig.SidebarState.AUTO,
  embedded: false,
  isFullScreen: false,
  setFullScreen: (value: boolean) => {},
  addScriptFinishedHandler: (func: () => void) => {},
  removeScriptFinishedHandler: (func: () => void) => {},
  activeTheme: baseTheme,
  setTheme: (theme: ThemeConfig) => {},
  availableThemes: [],
  addThemes: (themes: ThemeConfig[]) => {},
  sidebarChevronDownshift: 0,
  getBaseUriParts: getWindowBaseUriParts,
})
