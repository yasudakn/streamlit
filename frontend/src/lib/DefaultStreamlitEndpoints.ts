/**
 * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import axios, { AxiosRequestConfig, AxiosResponse, CancelToken } from "axios"
import { BaseUriParts, buildHttpUri } from "src/lib/UriUtil"
import { StreamlitEndpoints } from "src/lib/StreamlitEndpoints"
import { getCookie } from "./utils"

interface Props {
  getServerUri: () => BaseUriParts | undefined
  csrfEnabled: boolean
}

/** Default Streamlit server implementation of the StreamlitEndpoints interface. */
export class DefaultStreamlitEndpoints implements StreamlitEndpoints {
  private readonly getServerUri: () => BaseUriParts | undefined

  private readonly csrfEnabled: boolean

  private cachedServerUri?: BaseUriParts

  public constructor(props: Props) {
    this.getServerUri = props.getServerUri
    this.csrfEnabled = props.csrfEnabled
  }

  public buildComponentURL(componentName: string, path: string): string {
    return buildHttpUri(
      this.requireServerUri(),
      `component/${componentName}/${path}`
    )
  }

  public async uploadFileUploaderFile(
    file: File,
    widgetId: string,
    sessionId: string,
    onUploadProgress?: (progressEvent: any) => void,
    cancelToken?: CancelToken
  ): Promise<number> {
    const form = new FormData()
    form.append("sessionId", sessionId)
    form.append("widgetId", widgetId)
    form.append(file.name, file)

    return this.csrfRequest<number>("_stcore/upload_file", {
      cancelToken,
      method: "POST",
      data: form,
      responseType: "text",
      onUploadProgress,
    }).then(response => {
      // Sanity check. Axios should be returning a number here.
      if (typeof response.data === "number") {
        return response.data
      }

      throw new Error(
        `Bad uploadFile response: expected a number but got '${response.data}'`
      )
    })
  }

  public async fetchCachedForwardMsg(hash: string): Promise<Uint8Array> {
    const serverURI = this.requireServerUri()
    const rsp = await axios.request({
      url: buildHttpUri(serverURI, `_stcore/message?hash=${hash}`),
      method: "GET",
      responseType: "arraybuffer",
    })

    return new Uint8Array(rsp.data)
  }

  /**
   * Fetch the server URI. If our server is disconnected, default to the most
   * recent cached value of the URI. If we're disconnected and have no cached
   * value, throw an Error.
   */
  private requireServerUri(): BaseUriParts {
    const serverUri = this.getServerUri()
    if (serverUri != null) {
      this.cachedServerUri = serverUri
      return serverUri
    }

    if (this.cachedServerUri != null) {
      return this.cachedServerUri
    }

    throw new Error("not connected to a server!")
  }

  /**
   * Wrapper around axios.request to update the request config with
   * CSRF headers if client has CSRF protection enabled.
   */
  private csrfRequest<T = any, R = AxiosResponse<T>>(
    url: string,
    params: AxiosRequestConfig
  ): Promise<R> {
    const serverURI = this.requireServerUri()
    params.url = buildHttpUri(serverURI, url)

    if (this.csrfEnabled) {
      const xsrfCookie = getCookie("_xsrf")
      if (xsrfCookie != null) {
        params.headers = {
          "X-Xsrftoken": xsrfCookie,
          ...(params.headers || {}),
        }
        params.withCredentials = true
      }
    }

    return axios.request<T, R>(params)
  }
}
