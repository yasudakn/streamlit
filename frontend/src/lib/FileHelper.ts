/**
 * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { isFromWindows } from "src/lib/utils"

export enum FileSize {
  Gigabyte = "gb",
  Megabyte = "mb",
  Kilobyte = "kb",
  Byte = "b",
}

// There is a shift towards displaying storage in base 10 vs base 2
// but Windows is still displaying things in base 2. This does not handle
// all cases but for simplicity general rule is to use base 2 for Windows.
export const BYTE_CONVERSION_SIZE = isFromWindows() ? 1024 : 1000
const sizeUnitSequence = [
  FileSize.Gigabyte,
  FileSize.Megabyte,
  FileSize.Kilobyte,
  FileSize.Byte,
]

export const getSizeDisplay = (
  size: number,
  unit: FileSize,
  rounding = 1
): string => {
  if (!unit) {
    unit = FileSize.Byte
  }

  if (rounding < 0) {
    rounding = 0
  }

  if (size < 0) {
    throw new Error("Size must be greater than or equal to 0")
  }

  const sizeIndex = sizeUnitSequence.indexOf(unit)
  const nextUnitSize = size / BYTE_CONVERSION_SIZE
  if (sizeIndex && size > BYTE_CONVERSION_SIZE / 2) {
    return getSizeDisplay(
      nextUnitSize,
      sizeUnitSequence[sizeIndex - 1],
      rounding
    )
  }
  return `${size.toFixed(rounding)}${unit.toUpperCase()}`
}

export const sizeConverter = (
  size: number,
  inputUnit: FileSize,
  outputUnit: FileSize
): number => {
  if (size < 0) {
    throw Error("Size must be 0 or greater")
  }

  const inputLevel = sizeUnitSequence.findIndex(unit => unit === inputUnit)
  const outputLevel = sizeUnitSequence.findIndex(unit => unit === outputUnit)

  if (inputLevel === -1 || outputLevel === -1) {
    // Should not ever occur
    throw Error("Unexpected byte unit provided")
  }

  if (inputLevel === outputLevel) {
    return size
  }

  const levelsBetween = Math.abs(inputLevel - outputLevel)
  const byteDifference = BYTE_CONVERSION_SIZE ** levelsBetween

  if (inputLevel > outputLevel) {
    // Going from smaller to bigger
    return size / byteDifference
  }
  // Going from bigger to smaller
  return size * byteDifference
}
