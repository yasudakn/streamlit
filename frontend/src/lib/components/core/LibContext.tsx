/**
 * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from "react"

import { baseTheme, ThemeConfig } from "src/lib/theme"

export interface Props {
  /** True if the app is in full-screen mode. */
  isFullScreen: boolean

  /** Function that sets the `isFullScreen` property. */
  setFullScreen: (value: boolean) => void

  /**
   * Add a callback that will be called every time the app's script finishes
   * executing.
   */
  addScriptFinishedHandler: (func: () => void) => void

  /** Remove a previously-added scriptFinishedHandler callback. */
  removeScriptFinishedHandler: (func: () => void) => void

  /** The currently active app theme. */
  activeTheme: ThemeConfig

  /**
   * Set the app's active theme locally and send it the app's host (if any).
   * @see App.setAndSendTheme
   */
  setTheme: (theme: ThemeConfig) => void

  /** List of all available themes. */
  availableThemes: ThemeConfig[]

  /**
   * Call to add additional themes to the app.
   * @see ThemeCreatorDialog
   */
  addThemes: (themes: ThemeConfig[]) => void
}

export const LibContext = React.createContext<Props>({
  isFullScreen: false,
  setFullScreen: () => {},
  addScriptFinishedHandler: () => {},
  removeScriptFinishedHandler: () => {},
  activeTheme: baseTheme,
  setTheme: () => {},
  availableThemes: [],
  addThemes: () => {},
})
