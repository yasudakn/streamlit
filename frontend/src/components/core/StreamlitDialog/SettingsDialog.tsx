/**
 * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, {
  ChangeEvent,
  PureComponent,
  ReactElement,
  ReactNode,
} from "react"
import { ThemeConfig } from "src/theme"
import Button, { Kind } from "src/components/shared/Button"
import Modal, { ModalHeader, ModalBody } from "src/components/shared/Modal"
import AppContext, {
  Props as AppContextProps,
} from "src/components/core/AppContext"
import UISelectbox from "src/components/shared/Dropdown"
import { MetricsManager } from "src/lib/MetricsManager"

import {
  StyledCheckbox,
  StyledDialogBody,
  StyledFullRow,
  StyledHeader,
  StyledHr,
  StyledLabel,
  StyledSmall,
} from "./styled-components"
import { UserSettings } from "./UserSettings"

export interface Props {
  isServerConnected: boolean
  onClose: () => void
  onSave: (settings: UserSettings) => void
  settings: UserSettings
  allowRunOnSave: boolean
  developerMode: boolean
  openThemeCreator: () => void
  animateModal: boolean
}

/**
 * Implements a dialog that is used to configure user settings.
 */
export class SettingsDialog extends PureComponent<Props, UserSettings> {
  private activeSettings: UserSettings

  static contextType = AppContext

  constructor(props: Props) {
    super(props)
    // Holds the settings that will be saved when the "save" button is clicked.
    this.state = { ...this.props.settings }

    // Holds the actual settings that Streamlit is using.
    this.activeSettings = { ...this.props.settings }
  }

  private renderThemeCreatorButton = (): ReactElement | false =>
    this.props.developerMode && (
      <div>
        <Button onClick={this.props.openThemeCreator} kind={Kind.PRIMARY}>
          Edit active theme
        </Button>
      </div>
    )

  public render(): ReactNode {
    const themeIndex = this.context.availableThemes.findIndex(
      (theme: ThemeConfig) => theme.name === this.context.activeTheme.name
    )

    return (
      <Modal
        animate={this.props.animateModal}
        isOpen
        onClose={this.handleCancelButtonClick}
      >
        <ModalHeader>Settings</ModalHeader>
        <ModalBody>
          <StyledDialogBody>
            {this.props.allowRunOnSave && (
              <React.Fragment>
                <StyledFullRow>
                  <StyledHeader>Development</StyledHeader>
                  <label>
                    <StyledCheckbox
                      disabled={!this.props.isServerConnected}
                      type="checkbox"
                      name="runOnSave"
                      checked={
                        this.state.runOnSave && this.props.isServerConnected
                      }
                      onChange={this.handleCheckboxChange}
                    />{" "}
                    Run on save
                  </label>
                  <StyledSmall>
                    Automatically updates the app when the underlying code is
                    updated.
                  </StyledSmall>
                </StyledFullRow>

                <StyledFullRow>
                  <StyledHr />
                </StyledFullRow>
              </React.Fragment>
            )}

            <StyledFullRow>
              <StyledHeader>Appearance</StyledHeader>
              <label>
                <StyledCheckbox
                  type="checkbox"
                  name="wideMode"
                  checked={this.state.wideMode}
                  onChange={this.handleCheckboxChange}
                />{" "}
                Wide mode
              </label>
              <StyledSmall>
                Turn on to make this app occupy the entire width of the screen
              </StyledSmall>
            </StyledFullRow>

            {this.context.availableThemes.length && (
              <StyledFullRow>
                <StyledLabel>Theme</StyledLabel>
                <StyledSmall>Choose app and font colors/styles</StyledSmall>
                <UISelectbox
                  options={this.context.availableThemes.map(
                    (theme: ThemeConfig) => theme.name
                  )}
                  disabled={false}
                  onChange={this.handleThemeChange}
                  value={themeIndex}
                />
                {this.renderThemeCreatorButton()}
              </StyledFullRow>
            )}
          </StyledDialogBody>
        </ModalBody>
      </Modal>
    )
  }

  public componentDidMount(): void {
    this.setState({ ...this.activeSettings })
  }

  private changeSingleSetting = (name: string, value: boolean): void => {
    // TypeScript doesn't currently have a good solution for setState with
    // a dynamic key name:
    // https://github.com/DefinitelyTyped/DefinitelyTyped/issues/26635
    this.setState(state => ({ ...state, [name]: value }), this.saveSettings)
  }

  private handleCheckboxChange = (e: ChangeEvent<HTMLInputElement>): void => {
    this.changeSingleSetting(e.target.name, e.target.checked)
  }

  private handleThemeChange = (index: number): void => {
    const {
      activeTheme: oldTheme,
      availableThemes,
    }: AppContextProps = this.context
    const newTheme = availableThemes[index]

    MetricsManager.current.enqueue("themeChanged", {
      oldThemeName: oldTheme.name,
      newThemeName: newTheme.name,
    })

    this.context.setTheme(newTheme)
  }

  private handleCancelButtonClick = (): void => {
    // Discard settings from this.state by not saving them in this.settings.
    // this.settings = {...this.state};
    this.props.onClose()
  }

  private saveSettings = (): void => {
    this.activeSettings = { ...this.state }
    this.props.onSave(this.activeSettings)
  }
}
