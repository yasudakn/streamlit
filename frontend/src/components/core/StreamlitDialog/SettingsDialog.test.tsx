/**
 * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from "react"
import PropTypes from "prop-types"
import { createPresetThemes, lightTheme, darkTheme } from "src/theme"
import { shallow } from "src/lib/test_util"
import { Props as ContextProps } from "src/components/core/AppContext"
import UISelectbox from "src/components/shared/Dropdown"

import { SettingsDialog, Props } from "./SettingsDialog"

const mockSetTheme = jest.fn()
const mockAddThemes = jest.fn()

const getContext = (
  extend?: Partial<ContextProps>
): Partial<ContextProps> => ({
  activeTheme: lightTheme,
  setTheme: mockSetTheme,
  availableThemes: [],
  addThemes: mockAddThemes,
  ...extend,
})

// This is a workaround since enzyme does not support context yet
// https://github.com/enzymejs/enzyme/issues/2189
// @ts-ignore
SettingsDialog.contextTypes = {
  availableThemes: PropTypes.array,
  activeTheme: PropTypes.shape,
  setTheme: PropTypes.func,
  addThemes: PropTypes.func,
}

const getProps = (extend?: Partial<Props>): Props => ({
  isServerConnected: true,
  onClose: jest.fn(),
  onSave: jest.fn(),
  settings: { wideMode: false, runOnSave: false },
  allowRunOnSave: false,
  developerMode: true,
  animateModal: true,
  openThemeCreator: jest.fn(),
  ...extend,
})

describe("SettingsDialog", () => {
  it("renders without crashing", () => {
    const availableThemes = [lightTheme, darkTheme]
    const props = getProps()
    const context = getContext({ availableThemes })

    const wrapper = shallow(<SettingsDialog {...props} />, { context })

    expect(wrapper).toMatchSnapshot()
  })

  it("should render run on save checkbox", () => {
    const props = getProps({
      allowRunOnSave: true,
    })
    const context = getContext()
    const wrapper = shallow(<SettingsDialog {...props} />, { context })
    const checkboxes = wrapper.find("StyledCheckbox")

    expect(checkboxes).toHaveLength(2)
    expect(wrapper.state("runOnSave")).toBe(false)

    checkboxes
      .at(0)
      .simulate("change", { target: { name: "runOnSave", checked: true } })
    wrapper.update()

    expect(wrapper.state("runOnSave")).toBe(true)
    expect(props.onSave).toHaveBeenCalled()
    // @ts-ignore
    expect(props.onSave.mock.calls[0][0].runOnSave).toBe(true)
  })

  it("should render wide mode checkbox", () => {
    const props = getProps()
    const context = getContext()
    const wrapper = shallow(<SettingsDialog {...props} />, { context })
    const checkboxes = wrapper.find("StyledCheckbox")

    expect(checkboxes).toHaveLength(1)
    expect(wrapper.state("wideMode")).toBe(false)

    checkboxes
      .at(0)
      .simulate("change", { target: { name: "wideMode", checked: true } })
    wrapper.update()

    expect(wrapper.state("wideMode")).toBe(true)
    expect(props.onSave).toHaveBeenCalled()
    // @ts-ignore
    expect(props.onSave.mock.calls[0][0].wideMode).toBe(true)
  })

  it("should render theme selector", () => {
    const availableThemes = [lightTheme, darkTheme]
    const props = getProps()
    const context = getContext({ availableThemes })
    const wrapper = shallow(<SettingsDialog {...props} />, { context })
    const selectbox = wrapper.find(UISelectbox)
    const { options } = selectbox.props()

    expect(options).toHaveLength(2)
    expect(options).toEqual(availableThemes.map(theme => theme.name))

    selectbox.prop("onChange")(1)
    wrapper.update()
    expect(mockSetTheme).toHaveBeenCalled()
    // @ts-ignore
    expect(mockSetTheme.mock.calls[0][0]).toBe(darkTheme)
  })

  it("should show custom theme exists", () => {
    const presetThemes = createPresetThemes()
    const availableThemes = [...presetThemes, lightTheme]
    const props = getProps()
    const context = getContext({ availableThemes })
    const wrapper = shallow(<SettingsDialog {...props} />, { context })
    const selectbox = wrapper.find(UISelectbox)
    const { options } = selectbox.props()

    expect(options).toHaveLength(presetThemes.length + 1)
  })

  it("should show custom theme does not exists", () => {
    const presetThemes = createPresetThemes()
    const availableThemes = [...presetThemes]
    const props = getProps()
    const context = getContext({ availableThemes })
    const wrapper = shallow(<SettingsDialog {...props} />, { context })
    const selectbox = wrapper.find(UISelectbox)
    const { options } = selectbox.props()

    expect(options).toHaveLength(presetThemes.length)
  })

  it("should show theme creator button if in developer mode", () => {
    const availableThemes = [lightTheme, darkTheme]
    const props = getProps()
    const context = getContext({ availableThemes })
    const wrapper = shallow(<SettingsDialog {...props} />, { context })

    expect(wrapper.find("Button")).toHaveLength(1)
    expect(wrapper).toMatchSnapshot()
  })

  it("should call openThemeCreator if the button has been clicked", () => {
    const availableThemes = [...createPresetThemes()]
    const props = getProps()
    const context = getContext({ availableThemes })
    const wrapper = shallow(<SettingsDialog {...props} />, { context })
    wrapper.find("Button").simulate("click")

    expect(props.openThemeCreator).toHaveBeenCalled()
  })

  it("should hide the theme creator button if not in developer mode", () => {
    const availableThemes = [lightTheme, darkTheme]
    const props = getProps({ developerMode: false })
    const context = getContext({ availableThemes })
    const wrapper = shallow(<SettingsDialog {...props} />, { context })

    expect(wrapper.find("Button").exists()).toBe(false)
    expect(wrapper).toMatchSnapshot()
  })
})
