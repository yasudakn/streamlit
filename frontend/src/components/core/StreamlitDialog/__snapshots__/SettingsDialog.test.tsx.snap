// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsDialog renders without crashing 1`] = `
<Modal
  animate={true}
  isOpen={true}
  onClose={[Function]}
>
  <ModalHeader>
    Settings
  </ModalHeader>
  <ModalBody>
    <StyledDialogBody>
      <StyledFullRow>
        <StyledHeader>
          Appearance
        </StyledHeader>
        <label>
          <StyledCheckbox
            checked={false}
            name="wideMode"
            onChange={[Function]}
            type="checkbox"
          />
           
          Wide mode
        </label>
        <StyledSmall>
          Turn on to make this app occupy the entire width of the screen
        </StyledSmall>
      </StyledFullRow>
      <StyledFullRow>
        <StyledLabel>
          Theme
        </StyledLabel>
        <StyledSmall>
          Choose app and font colors/styles
        </StyledSmall>
        <Selectbox
          disabled={false}
          onChange={[Function]}
          options={
            Array [
              "Light",
              "Dark",
            ]
          }
          value={0}
        />
        <div>
          <Button
            kind="secondary"
            onClick={[MockFunction]}
          >
            Edit active theme
          </Button>
        </div>
      </StyledFullRow>
    </StyledDialogBody>
  </ModalBody>
</Modal>
`;

exports[`SettingsDialog should hide the theme creator button if not in developer mode 1`] = `
<Modal
  animate={true}
  isOpen={true}
  onClose={[Function]}
>
  <ModalHeader>
    Settings
  </ModalHeader>
  <ModalBody>
    <StyledDialogBody>
      <StyledFullRow>
        <StyledHeader>
          Appearance
        </StyledHeader>
        <label>
          <StyledCheckbox
            checked={false}
            name="wideMode"
            onChange={[Function]}
            type="checkbox"
          />
           
          Wide mode
        </label>
        <StyledSmall>
          Turn on to make this app occupy the entire width of the screen
        </StyledSmall>
      </StyledFullRow>
      <StyledFullRow>
        <StyledLabel>
          Theme
        </StyledLabel>
        <StyledSmall>
          Choose app and font colors/styles
        </StyledSmall>
        <Selectbox
          disabled={false}
          onChange={[Function]}
          options={
            Array [
              "Light",
              "Dark",
            ]
          }
          value={0}
        />
      </StyledFullRow>
    </StyledDialogBody>
  </ModalBody>
</Modal>
`;

exports[`SettingsDialog should show theme creator button if in developer mode 1`] = `
<Modal
  animate={true}
  isOpen={true}
  onClose={[Function]}
>
  <ModalHeader>
    Settings
  </ModalHeader>
  <ModalBody>
    <StyledDialogBody>
      <StyledFullRow>
        <StyledHeader>
          Appearance
        </StyledHeader>
        <label>
          <StyledCheckbox
            checked={false}
            name="wideMode"
            onChange={[Function]}
            type="checkbox"
          />
           
          Wide mode
        </label>
        <StyledSmall>
          Turn on to make this app occupy the entire width of the screen
        </StyledSmall>
      </StyledFullRow>
      <StyledFullRow>
        <StyledLabel>
          Theme
        </StyledLabel>
        <StyledSmall>
          Choose app and font colors/styles
        </StyledSmall>
        <Selectbox
          disabled={false}
          onChange={[Function]}
          options={
            Array [
              "Light",
              "Dark",
            ]
          }
          value={0}
        />
        <div>
          <Button
            kind="secondary"
            onClick={[MockFunction]}
          >
            Edit active theme
          </Button>
        </div>
      </StyledFullRow>
    </StyledDialogBody>
  </ModalBody>
</Modal>
`;
